boot-qemu.py

```diff
@@ -0,0 +1,247 @@
+        usable_cpus = os.cpu_count()
+        return min(usable_cpus, config_nr_cpus)
+
+    def _get_qemu_ver_string(self):
+        if not self._qemu_path:
+            raise RuntimeError('No path to QEMU set?')
+        qemu_ver = subprocess.run([self._qemu_path, '--version'],
+                                  capture_output=True,
+                                  check=True,
+                                  text=True)
+        return qemu_ver.stdout.splitlines()[0]
+
+    def _have_dev_kvm_access(self):
+        return os.access('/dev/kvm', os.R_OK | os.W_OK)
+
+    def _prepare_initrd(self):
+        if not self._initrd_arch:
+            raise RuntimeError('No initrd architecture specified?')
+        if not (src := Path(BOOT_UTILS, 'images', self._initrd_arch,
+                            'rootfs.cpio.zst')):
+            raise FileNotFoundError(f"initrd ('{src}') does not exist?")
+
+        (dst := src.with_suffix('')).unlink(missing_ok=True)
+
+        utils.check_cmd('zstd')
+        subprocess.run(['zstd', '-d', src, '-o', dst, '-q'], check=True)
+
+        return dst
+
+    def run(self):
+        # Make sure QEMU binary is configured and available
+        if not self._qemu_arch:
+            raise RuntimeError('No QEMU architecture set?')
+        qemu_bin = f"qemu-system-{self._qemu_arch}"
+        if not (qemu_path := shutil.which(qemu_bin)):
+            raise RuntimeError(
+                f'{qemu_bin} could not be found on your system?')
+        self._qemu_path = Path(qemu_path)
+
+        # Locate kernel if it was not specified
+        if self.kernel:
+            if not self.kernel_dir:
+                self.kernel_dir = self.kernel.parent
+        else:
+            if not self.kernel_dir:
+                raise RuntimeError(
+                    'No kernel image or kernel build folder specified?')
+            if not self._default_kernel_path:
+                raise RuntimeError('No default kernel path specified?')
+
+            possible_kernel_locations = {
+                Path(self.kernel_dir,
+                     self._default_kernel_path),  # default (kbuild)
+                Path(self.kernel_dir,
+                     self._default_kernel_path.name),  # tuxmake
+            }
+            for loc in possible_kernel_locations:
+                if loc.exists():
+                    self.kernel = loc
+                    break
+            if not self.kernel:
+                possible_locations = "', '".join(
+                    str(path) for path in possible_kernel_locations)
+                raise FileNotFoundError(
+                    f"{self._default_kernel_path.name} could not be found at possible locations ('{possible_locations}')",
+                )
+
+        # Kernel options
+        if self.cmdline:
+            self._qemu_args += ['-append', ' '.join(self.cmdline)]
+        self._qemu_args += ['-kernel', self.kernel]
+        self._qemu_args += ['-initrd', self._prepare_initrd()]
```

Comment:
if `run` is called more than once, we're going to keep appending to `self._qemu_args`. Do we expect `run` to be called once?

Reply:
No, `run()` should only be called once.